presto-main/src/main/java/com/facebook/presto/operator/BlockEncodingBuffers.java

```diff
@@ -0,0 +1,432 @@
+import io.airlift.slice.SliceOutput;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+
+import static com.facebook.presto.array.Arrays.ExpansionFactor.LARGE;
+import static com.facebook.presto.array.Arrays.ExpansionFactor.SMALL;
+import static com.facebook.presto.array.Arrays.ensureCapacity;
+import static com.facebook.presto.operator.MoreByteArrays.fill;
+import static com.facebook.presto.operator.UncheckedByteArrays.setByteUnchecked;
+import static com.google.common.base.Verify.verify;
+import static io.airlift.slice.SizeOf.SIZE_OF_BYTE;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public abstract class BlockEncodingBuffers
+{
+    private static final int BITS_IN_BYTE = 8;
+
+    // The block after peeling off the Dictionary or RLE wrappings.
+    protected Block decodedBlock;
+
+    // The number of positions (rows) to be copied from the incoming block.
+    protected int positionCount;
+
+    // The number of positions (rows) to be copied from the incoming block within current batch.
+    protected int batchSize;
+
+    // The offset into positions/mappedPositions array that the current batch starts copying.
+    // I.e. in each batch we copy the values of rows from positions[positionsOffset] to positions[positionsOffset + batchSize]
+    protected int positionsOffset;
+
+    // The number of positions (rows) buffered so far
+    protected int bufferedPositionCount;
+
+    // Whether the positions array was mapped to mappedPositions
+    protected boolean positionsMapped;
+
+    // The positions (row numbers) of the incoming Block to be copied to this buffer.
+    // All top level BlockEncodingBuffers for the same partition share the same positions array.
+    private int[] positions;
+
+    // The mapped positions of the incoming Dictionary or RLE Block
+    @Nullable
+    private int[] mappedPositions;
+
+    @Nullable
+    private byte[] nullsBuffer;
+
+    // The address offset in the nullsBuffer if new values are to be added.
+    private int nullsBufferIndex;
+
+    // For each batch we put the nulls values up to a multiple of 8 into nullsBuffer. The rest is kept in remainingNulls.
+    private final boolean[] remainingNulls = new boolean[BITS_IN_BYTE];
+
+    // Number of null values not put into nullsBuffer from last batch
+    private int remainingNullsCount;
+
+    // Boolean indicating whether there are any null values in the nullsBuffer. It is possible that all values are non-null.
+    private boolean hasEncodedNulls;
+
+    public static BlockEncodingBuffers createBlockEncodingBuffers(DecodedBlockNode decodedBlockNode)
+    {
+        requireNonNull(decodedBlockNode, "decodedBlockNode is null");
+
+        Object decodedBlock = decodedBlockNode.getDecodedBlock();
+
+        // Skip the Dictionary/Rle block node. The mapping info is not needed when creating buffers.
```

Comment:
Curious why not do "position mapping" for Dictionary/RLE block here ? This makes the code a bit difficult to follow though. 

Maybe explain why not doing "position mapping" here in the comment, and refer to the methods doing remapping (e.g. `Dictionary/RLE block index remapping is done in setupDecodedBlocksAndPositions method, blabla`.

Reply:
@wenleix does the following comments answer your question?
```

        // Skip the Dictionary/Rle block node. The mapping info is not needed when creating buffers.
        // This is because the BlockEncodingBuffers is only created once, while position mapping for Dictionar/Rle blocks
        // need to be done for every incoming block.
```

Reply:
@yingsu00 : This nice comment answers my questions. Thanks for the explanation!